src/cycle.rs

```diff
@@ -121,36 +133,39 @@ impl std::iter::Extend<DatabaseKeyIndex> for CycleHeads {
     fn extend<T: IntoIterator<Item = DatabaseKeyIndex>>(&mut self, iter: T) {
         let mut iter = iter.into_iter();
         if let Some(first) = iter.next() {
-            let heads = self.0.get_or_insert(Box::new(FxHashSet::default()));
-            heads.insert(first);
-            heads.extend(iter)
+            let heads = &mut **self.0.get_or_insert_with(|| Box::new(Vec::new()));
+            if !heads.contains(&first) {
+                heads.push(first);
+            }
+            for head in iter {
+                if !heads.contains(&head) {
+                    heads.push(head);
+                }
+            }
         }
     }
 }
 
-impl std::iter::IntoIterator for CycleHeads {
+impl IntoIterator for CycleHeads {
     type Item = DatabaseKeyIndex;
-    type IntoIter = std::collections::hash_set::IntoIter<Self::Item>;
+    type IntoIter = <Vec<Self::Item> as IntoIterator>::IntoIter;
 
     fn into_iter(self) -> Self::IntoIter {
         self.0.map(|heads| *heads).unwrap_or_default().into_iter()
     }
 }
 
-// This type can be removed once MSRV is 1.83+ and we have Default for hashset iterators.
-pub(crate) struct CycleHeadsIter<'a>(
-    Option<std::collections::hash_set::Iter<'a, DatabaseKeyIndex>>,
-);
+pub struct CycleHeadsIter<'a>(std::slice::Iter<'a, DatabaseKeyIndex>);
```

Comment:
I don't know if you consider this an improvement, but we can eliminate `CycleHeadsIter` and implement `IntoIterator` for `&'a CycleHeads` like so:

```rs
impl<'a> std::iter::IntoIterator for &'a CycleHeads {
    type Item = DatabaseKeyIndex;
    type IntoIter = std::iter::Copied<std::slice::Iter<'a, DatabaseKeyIndex>>;

    fn into_iter(self) -> Self::IntoIter {
        self.0
            .as_ref()
            .map(|heads| heads.iter())
            .unwrap_or_default()
            .copied()
    }
}
```

Reply:
This causes `new[InternedInput]` to regress due to a `report_tracked_read` becoming more expensive, so something optimizes differently there 😕 Will drop our `Extend` impl as we can optimize around knowing the concrete iterator anyways.